Fix ComputeTimeSmart test failure with -DDEBUG_LOCKORDER

ryanofsky opened the pull request:
Failure looks like:

```
Entering test case "ComputeTimeSmart"
test_bitcoin: sync.cpp:100: void potential_deadlock_detected(const std::pair<void*, void*>&, const LockStack&, const LockStack&): Assertion `false' failed.
unknown location(0): fatal error in "ComputeTimeSmart": signal: SIGABRT (application abort requested)
wallet/test/wallet_tests.cpp(566): last checkpoint
```

Reproducible with:

```
./configure --enable-debug
make -C src test/test_bitcoin && src/test/test_bitcoin --log_level=test_suite --run_test=wallet_tests/ComputeTimeSmart
```

Seems to be caused by acquiring `cs_main` inside `CWallet::ComputeTimeSmart` in #11041.

I think this may be causing timeouts on travis like: https://travis-ci.org/bitcoin/bitcoin/jobs/353005676#L2692

ryanofsky committed:
Fix ComputeTimeSmart test failure with -DDEBUG_LOCKORDER

Failure looks like:

    Entering test case "ComputeTimeSmart"
    test_bitcoin: sync.cpp:100: void potential_deadlock_detected(const std::pair<void*, void*>&, const LockStack&, const LockStack&): Assertion `false' failed.
    unknown location(0): fatal error in "ComputeTimeSmart": signal: SIGABRT (application abort requested)
    wallet/test/wallet_tests.cpp(566): last checkpoint

Reproducible with:

    ./configure --enable-debug
    make -C src test/test_bitcoin && src/test/test_bitcoin --log_level=test_suite --run_test=wallet_tests/ComputeTimeSmart

Happens due to "92fabcd443 Add LookupBlockIndex function" which acquires
cs_main from inside CWallet::ComputeTimeSmart.

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -899,6 +899,7 @@ bool CWallet::MarkReplaced(const uint256& originalHash, const uint256& newHash)
 
 bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
 {
+    AssertLockHeld(cs_main); // Needed for ComputeTimeSmart
```
Wouldn't it make sense to instead move this down to ComputeTimeSmart?

```diff
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -899,7 +899,6 @@ bool CWallet::MarkReplaced(const uint256& originalHash, const uint256& newHash)
 
 bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
 {
-    AssertLockHeld(cs_main); // Needed for ComputeTimeSmart
     LOCK(cs_wallet);
 
     CWalletDB walletdb(*dbw, "r+", fFlushOnClose);
@@ -3815,10 +3814,8 @@ unsigned int CWallet::ComputeTimeSmart(const CWalletTx& wtx) const
     unsigned int nTimeSmart = wtx.nTimeReceived;
     if (!wtx.hashUnset()) {
         const CBlockIndex* pindex = nullptr;
-        {
-            LOCK(cs_main);
-            pindex = LookupBlockIndex(wtx.hashBlock);
-        }
+        AssertLockHeld(cs_main);
+        pindex = LookupBlockIndex(wtx.hashBlock);
         if (pindex) {
             int64_t latestNow = wtx.nTimeReceived;
             int64_t latestEntry = 0;

promag commented:
Actually I think it's enough to fix the test and remove the lock. The lock assertion is already in LookupBlockIndex.

ryanofsky commented:
> Actually I think it's enough to fix the test and remove the lock. The lock assertion is already in LookupBlockIndex.

Sounds good. Removed the lock and new assert.

Updated 84ae51a95c40e1e11520f41c1f8783504de694b3 -> 33eb9071b98e63fd66c1211254e420535320eebd ([pr/locksmart.1](https://github.com/ryanofsky/bitcoin/commits/pr/locksmart.1) -> [pr/locksmart.2](https://github.com/ryanofsky/bitcoin/commits/pr/locksmart.2))

promag commented:
utACK 33eb9071.

Sorry for the bug, thanks @ryanofsky for the quick fix.

MarcoFalke commented:
utACK 33eb9071b98e63fd66c1211254e420535320eebd
